### PR TITLE
#837 Enhance Callback method

### LINKED ISSUE #837: Enhance Callback When Person Lookup Fails
## Value Statement

**As** A business / service
**I want to** get callbacks when a person lookup fails via MPI/VA Profile
**So that** I can keep all my callbacks in one place without having to call /v2/notifications/\<notification_id\>

## Checklist
- [ ] Callback method called when a failure occurs during the lookup process:
- MPI: Deceased
- MPI: No VA Profile ID
- VA Profile: No contact info found
- VA Profile: Preference declined
- [ ] Unit Tests

## Additional Info/Resources

- [Callback method](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/celery/service_callback_tasks.py#L223)
- Most likely [injection point](https://github.com/department-of-veterans-affairs/notification-api/blob/master/app/celery/lookup_va_profile_id_task.py#L60) for the method
- Test [users](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/Administrative/vagov-users/mvi-staging-users.csv)

### PR DESCRIPTION
**Pull request by trevor2718**
# Description

Callback method called when a failure occurs during the lookup process.

Fixes #837 (Enhance Callback When Person Lookup Fails)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] All unit tests pass

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kalbfled** [comment]
Regarding how to test this change, I might make an API call that I expect to execute the lookup_va_profile_id task and raise IdentifierNotFound in that task.  If I correctly understand the intended behavior, the exception handler should update a Notification instance in the database to reflect a permanent failure condition.  You could query the database in your test, after pausing for a second, and make sure the Notification instance reflects the update.

**Comment by trevor2718** [comment]
> Regarding how to test this change, I might make an API call that I expect to execute the lookup_va_profile_id task and raise IdentifierNotFound in that task. If I correctly understand the intended behavior, the exception handler should update a Notification instance in the database to reflect a permanent failure condition. You could query the database in your test, after pausing for a second, and make sure the Notification instance reflects the update.

@kalbfled The method check_and_queue_callback_task already has tests written and it checks the existence of the service_callback_api only so I didn't write a test to check the database record. If you think that test case is necessary, I can write one for sure.

**Comment by kalbfled** [comment]
In that case, I would write a test that mocks the call to check_and_queue_callback_task and uses [assert_called_once_with](https://docs.python.org/3.8/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with) to ensure the function is called as expected.

**Comment by EvanParish** [comment]
> In that case, I would write a test that mocks the call to check_and_queue_callback_task and uses [assert_called_once_with](https://docs.python.org/3.8/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with) to ensure the function is called as expected.

@trevor2718 this is what I was talking about when we spoke about unit testing. I probably didn't make that very clear, sorry about that, but I agree with Dave here. There should probably be a test in `tests/app/celery/test_lookup_va_profile_id_task.py` that makes sure the function is called as expected.

**Review by EvanParish — APPROVED** [review]
Looks good

File: tests/app/celery/test_lookup_va_profile_id_task.py [hunk 1]
```diff
@@ -269,3 +269,48 @@ def test_caught_exceptions_should_set_status_reason_on_notification(
...
         notification.id, notification_status, status_reason=failure_reason
     )
+
+
+@pytest.mark.parametrize(
+    "exception, reason",
+    [
+        (BeneficiaryDeceasedException('some error'), BeneficiaryDeceasedException.failure_reason),
+        (IdentifierNotFound('some error'), IdentifierNotFound.failure_reason),
+        (MultipleActiveVaProfileIdsException('some error'), MultipleActiveVaProfileIdsException.failure_reason)
```
Comment (kalbfled) [inline tests/app/celery/test_lookup_va_profile_id_task.py:279]: There are 6 exceptions that will execute the code you are testing, but you've only parametrized 3 of them.  Is there a reason you omitted the other 3?

**Review by kalbfled — CHANGES_REQUESTED** [review]
In addition to my comment on the code, I think we also need some negative tests.  i.e.  MpiRetryableException should not results in the callback being called.

File: tests/app/celery/test_lookup_va_profile_id_task.py (hunk 1 above)
Comment (trevor2718) [reply tests/app/celery/test_lookup_va_profile_id_task.py:279]: Sorry about not covering all the test cases, I totally forgot. (Maybe I was too excited about fixing the problem we had). I will write the additional tests today and test them out.

**Review by EvanParish — APPROVED** [review]

**Review by kalbfled — APPROVED** [review]

**Pull request merged**